cranelift/wasm/src/code_translator.rs

```diff
@@ -3556,6 +3564,205 @@ fn tag_memory_region<FE>(
+
+    let size = builder.block_params(st2g_loop_condition_block)[0];
+    let iter_ptr = builder.block_params(st2g_loop_condition_block)[1];
+
+    let cond = builder.ins().icmp_imm(IntCC::UnsignedGreaterThan, size, 32);
+    canonicalise_brif(
+        builder,
+        cond,
+        st2g_loop_body_block,
+        &[size, iter_ptr],
+        last_stg_condition_block,
+        &[size, iter_ptr],
+    );
+
+    // === st2g loop body block
+    builder.switch_to_block(st2g_loop_body_block);
+
+    let size_counter = builder.block_params(st2g_loop_body_block)[0];
+    let iter_ptr = builder.block_params(st2g_loop_body_block)[1];
+
+    builder.ins().arm64_st2g(tagged_ptr, iter_ptr);
+
+    let size_counter = builder.ins().iadd_imm(size_counter, -32);
+    // TODO: moving this const out of the loop might be an optimization possibility
+    // iter_ptr += 32, but trap on overflow
+    let offset = builder.ins().iconst(I64, 32);
+    let iter_ptr =
+        builder
+            .ins()
+            .uadd_overflow_trap(iter_ptr, offset, ir::TrapCode::HeapOutOfBounds);
+
+    builder.ins().jump(st2g_loop_condition_block, &[size_counter, iter_ptr]);
+
+    builder.seal_block(st2g_loop_body_block);
+    builder.seal_block(st2g_loop_condition_block);
+
+    // === last stg condition block
+    builder.switch_to_block(last_stg_condition_block);
+
+    let size = builder.block_params(last_stg_condition_block)[0];
+    let iter_ptr = builder.block_params(last_stg_condition_block)[1];
+
+    let cond = builder.ins().icmp_imm(IntCC::NotEqual, size, 0);
+    canonicalise_brif(
+        builder,
+        cond,
+        last_stg_body_block,
+        &[size, iter_ptr],
+        next_block,
+        &[],
+    );
+
+    builder.seal_block(last_stg_condition_block);
+
+    // === last stg body block
+    builder.switch_to_block(last_stg_body_block);
+
+    let iter_ptr = builder.block_params(last_stg_body_block)[1];
+
+    builder.ins().arm64_stg(tagged_ptr, iter_ptr);
+
+    // TODO: can I just remove the final size and iter_ptr calculations, since they aren't used afterwards anymore?
+
+    builder.ins().jump(next_block, &[]);
+
+    builder.seal_block(last_stg_body_block);
+
+    // === Next block
+    builder.switch_to_block(next_block);
+
+    builder.seal_block(next_block);
+
+    Ok(())
+}
+
+// TODO: implement as well for future performance comparisons
+fn tag_memory_region_optimized_v2<FE>(
+    iter_ptr: Value,
+    tagged_ptr: Value,
+    size: Value,
+    builder: &mut FunctionBuilder,
+    environ: &mut FE,
+) -> WasmResult<()>
+where
+    FE: FuncEnvironment + ?Sized,
+{
+    // Pseudo code (high level):
+    //
+    // assert_16_byte_aligned(size)
+    //
+    // if (size % 32 == 0)
+    //     for (size; size != 0; size -= 32, iter_ptr += 32):
+    //         st2g(tagged_ptr, iter_ptr)
+    // else
+    //     for (size; size != 0; size -= 16, iter_ptr += 16):
+    //         stg(tagged_ptr, iter_ptr)
```

Comment:
Looks good! I guess for smallish `n`s this might be faster than the following approach, but for larger ones (let's say `n = 2 << 30 + 16` = 1 GiB + 16 bytes) this might be faster:

```
for (size; size >= 32; size -= 32, iter_ptr += 32):
  st2g(tagged_ptr, iter_ptr)

if (size >= 16):
   stg(tagged_ptr, iter_ptr)
```

Reply:
I implemented the approach you described already, the code snippet you pointed out is just a second potential optimization/version that I have only outlined and not implemented yet.

Reply:
Sorry --  did not see that